mlflow/shap.py

```diff
@@ -179,3 +257,323 @@ def predict_function(X) -> pred:
+                          base64-encoded.
+    """
+    import shap
+
+    if os.path.exists(path):
+        raise MlflowException(
+            message="Path '{}' already exists".format(path), error_code=RESOURCE_ALREADY_EXISTS,
+        )
+
+    os.makedirs(path)
+    if mlflow_model is None:
+        mlflow_model = Model()
+    if signature is not None:
+        mlflow_model.signature = signature
+    if input_example is not None:
+        _save_example(mlflow_model, input_example, path)
+
+    underlying_model_flavor = None
+    underlying_model_path = None
+
+    # saving the underlying model if required
+    if serialize_model_using_mlflow:
+        underlying_model_flavor = get_underlying_model_flavor(explainer.model)
+
+        if underlying_model_flavor != _UNKNOWN_MODEL_FLAVOR:
+            explainer.model.save = None
+            underlying_model_path = os.path.join(path, _UNDERLYING_MODEL_SUBPATH)
+
+        if underlying_model_flavor == "sklearn":
+            mlflow.sklearn.save_model(explainer.model.model.__self__, underlying_model_path)
+        elif underlying_model_flavor == "pytorch":
+            mlflow.pytorch.save_model(explainer.model.model, underlying_model_path)
+
+    # saving the explainer object
+    explainer_data_subpath = "explainer.shap"
+    explainer_output_path = os.path.join(path, explainer_data_subpath)
+    with open(explainer_output_path, "wb") as explainer_output_file_handle:
+        explainer.save(explainer_output_file_handle)
+
+    conda_env_subpath = "conda.yaml"
+    if conda_env is None:
+        conda_env = get_default_conda_env()
+    elif not isinstance(conda_env, dict):
+        with open(conda_env, "r") as f:
+            conda_env = yaml.safe_load(f)
+
+    # merging the conda environment generated by serializing the underlying model
+    if underlying_model_path is not None:
+        underlying_model_conda_path = os.path.join(underlying_model_path, "conda.yaml")
+        with open(underlying_model_conda_path, "r") as underlying_model_conda_file:
+            underlying_model_conda_env = yaml.safe_load(underlying_model_conda_file)
+        conda_env = _merge_environments(conda_env, underlying_model_conda_env)
+
+    with open(os.path.join(path, conda_env_subpath), "w") as f:
+        yaml.safe_dump(conda_env, stream=f, default_flow_style=False)
+
+    pyfunc.add_to_model(
+        mlflow_model,
+        loader_module="mlflow.shap",
+        model_path=explainer_data_subpath,
+        underlying_model_flavor=underlying_model_flavor,
+        env=conda_env_subpath,
+    )
+
+    mlflow_model.add_flavor(
+        FLAVOR_NAME,
+        shap_version=shap.__version__,
+        serialized_explainer=explainer_data_subpath,
+        underlying_model_flavor=underlying_model_flavor,
+    )
+
+    mlflow_model.save(os.path.join(path, MLMODEL_FILE_NAME))
+
+
+def _merge_environments(shap_environment, model_environment):
```

Comment:
Can we simplify this function by introducing a helper function like below?

```python
def _get_conda_and_pip_dependencies(conda_env):
    ...
    return conda_deps, pip_deps
```

With this function, `_merge_environments` would look like:

```python
def _merge_environments(...):
    ...
    shap_conda_deps, shap_pip_deps = _get_conda_and_pip_dependencies(shap_environment)
    model_conda_deps, shap_pip_deps = _get_conda_and_pip_dependencies(shap_environment)

    merged_conda_deps = set(shap_conda_deps + model_conda_deps)
    merged_pip_deps = set(shap_pip_deps + model_pip_deps)
    ...
```

Reply:
Added helper function